File: engine/argot/extract.py
from __future__ import annotations

import argparse
import json
import sys
from dataclasses import asdict
from pathlib import Path

import pygit2

from argot.dataset import HunkRecord
from argot.git_walk import walk_repo
from argot.tokenize import language_for_path

CONTEXT_LINES = 50


def _extract_context(
    source_lines: list[str],
    hunk_start: int,
    hunk_end: int,
) -> tuple[list[str], list[str], list[str]]:
    before_start = max(0, hunk_start - CONTEXT_LINES)
    after_end = min(len(source_lines), hunk_end + CONTEXT_LINES)
    return (
        source_lines[before_start:hunk_start],
        source_lines[hunk_start:hunk_end],
        source_lines[hunk_end:after_end],
    )


def main() -> None:
    parser = argparse.ArgumentParser(description="Extract dataset from git history")
    parser.add_argument("repo_path", help="Path to git repository")
    parser.add_argument("--out", default=".argot/dataset.jsonl", help="Output JSONL path")
    parser.add_argument("--limit", type=int, default=None, help="Max number of records to emit")
    args = parser.parse_args()

    repo_path = args.repo_path
    out_path = Path(args.out)

    try:
        pygit2.Repository(repo_path)
    except pygit2.GitError:
        print(f"error: repository not found at {repo_path!r}", file=sys.stderr)
        sys.exit(2)

    out_path.parent.mkdir(parents=True, exist_ok=True)

    count = 0

    with open(out_path, "w") as fh:
        for commit, file_path, post_blob, hunks in walk_repo(repo_path):
            lang = language_for_path(file_path)
            if lang is None:
                continue

            try:
                source_lines = post_blob.decode("utf-8", errors="replace").splitlines()
            except Exception:
                continue

            parent_sha = str(commit.parents[0].id) if commit.parents else None
            author_date_iso = commit.author.time

            for hunk in hunks:
                hunk_start = hunk.new_start - 1  # convert to 0-indexed
                hunk_end = hunk_start + hunk.new_lines

                if hunk_start < 0 or hunk_end > len(source_lines):
                    continue

                ctx_before_lines, hunk_lines, ctx_after_lines = _extract_context(
                    source_lines, hunk_start, hunk_end
                )

                before_start_abs = max(0, hunk_start - CONTEXT_LINES)
                after_start_abs = hunk_end

                from argot.tokenize import tokenize_lines

                context_before = tokenize_lines(source_lines, lang, before_start_abs, hunk_start)
                hunk_tokens = tokenize_lines(source_lines, lang, hunk_start, hunk_end)
                context_after = tokenize_lines(
                    source_lines,
                    lang,
                    after_start_abs,
                    min(len(source_lines), after_start_abs + CONTEXT_LINES),
                )

                record = HunkRecord(
                    commit_sha=str(commit.id),
                    file_path=file_path,
                    language=lang,
                    hunk_start_line=hunk_start,
                    hunk_end_line=hunk_end,
                    context_before=context_before,
                    hunk_tokens=hunk_tokens,
                    context_after=context_after,
                    parent_sha=parent_sha,
                    author_date_iso=str(author_date_iso),
                )

                fh.write(json.dumps(asdict(record)))
                fh.write("\n")
                count += 1

                if args.limit is not None and count >= args.limit:
                    print(f"Reached limit of {args.limit} records", file=sys.stderr)
                    print(f"Wrote {count} records to {out_path}")
                    return

    if count == 0:
        print("error: no hunks found — repository may have no history", file=sys.stderr)
        sys.exit(2)

    print(f"Wrote {count} records to {out_path}")


if __name__ == "__main__":
    main()

File: crates/argot-rules-voice/src/extract.rs
//! Dataset extraction.
//!
//! Walks git history, tokenises each changed hunk plus up to
//! [`CONTEXT_LINES`] of surrounding context, and writes one JSON
//! [`HunkRecord`](argot_lang::dataset::HunkRecord) per line.

use anyhow::Result;
use argot_engine::git_walk::{open_repo, resolve_shas, walk_commits, walk_repo, WalkItem};
use argot_lang::dataset::HunkRecord;
use argot_lang::text::splitlines;
use argot_lang::tokenize::{language_for_path_ctx, tokenize_lines};
use std::collections::HashSet;
use std::io::Write;
use std::ops::ControlFlow;

/// Lines of context captured on each side of a hunk.
pub const CONTEXT_LINES: usize = 50;

/// Extraction errors that map to specific CLI exit codes / messages.
#[derive(Debug, thiserror::Error)]
pub enum ExtractError {
    #[error("no commits found for ref {0:?}")]
    NoCommitsForRef(String),
}

/// Result of a completed extraction.
#[derive(Debug, Clone, Copy)]
pub struct ExtractStats {
    pub count: usize,
    pub limit_reached: bool,
}

/// Extract records from `repo_path` (optionally scoped to `reference`) and
/// write them as JSONL to `out`. Returns record counts.
///
/// Topological walk, single-parent commits only, hunk bounds clamped to the
/// post-image, context of [`CONTEXT_LINES`] each side.
pub fn write_dataset<W: Write>(
    repo_path: &str,
    reference: Option<&str>,
    limit: Option<usize>,
    out: &mut W,
) -> Result<ExtractStats> {
    let filter: Option<HashSet<String>> = match reference {
        Some(r) if !r.is_empty() => {
            let repo = open_repo(repo_path)?;
            let shas = resolve_shas(&repo, r).unwrap_or_default();
            if shas.is_empty() {
                return Err(ExtractError::NoCommitsForRef(r.to_string()).into());
            }
            Some(shas)
        }
        _ => None,
    };

    let mut count = 0usize;
    let mut limit_reached = false;

    // `.h` and `.inc` route by what the repo itself writes — decided once here
    // so the dataset labels match how calibrate/check route.
    let repo_langs = argot_engine::corpus::repo_langs(std::path::Path::new(repo_path));

    {
        let mut visit = |item: WalkItem| -> Result<ControlFlow<()>> {
            let lang = match language_for_path_ctx(&item.file_path, repo_langs) {
                Some(l) => l,
                None => return Ok(ControlFlow::Continue(())),
            };
            let source = String::from_utf8_lossy(&item.post_blob);
            let lines = splitlines(&source);
            let nlines = lines.len() as i64;

            for hunk in &item.hunks {
                let hunk_start = hunk.new_start as i64 - 1; // 0-indexed
                let hunk_end = hunk_start + hunk.new_lines as i64;
                if hunk_start < 0 || hunk_end > nlines {
                    continue;
                }
                let hunk_start = hunk_start as usize;
                let hunk_end = hunk_end as usize;

                let before_start = hunk_start.saturating_sub(CONTEXT_LINES);
                let after_start = hunk_end;

                let context_before = tokenize_lines(&lines, lang, before_start, hunk_start);
                let hunk_tokens = tokenize_lines(&lines, lang, hunk_start, hunk_end);
                let context_after = tokenize_lines(
                    &lines,
                    lang,
                    after_start,
                    (after_start + CONTEXT_LINES).min(lines.len()),
                );

                let record = HunkRecord {
                    commit_sha: item.commit_id.clone(),
                    file_path: item.file_path.clone(),
                    language: lang,
                    hunk_start_line: hunk_start,
                    hunk_end_line: hunk_end,
                    context_before,
                    hunk_tokens,
                    context_after,
                    parent_sha: item.parent_id.clone(),
                    author_date_iso: item.author_time.to_string(),
                };

                out.write_all(argot_engine::json::to_string(&record)?.as_bytes())?;
                out.write_all(b"\n")?;
                count += 1;

                if let Some(l) = limit {
                    if count >= l {
                        limit_reached = true;
                        break;
                    }
                }
            }

            if limit_reached {
                Ok(ControlFlow::Break(()))
            } else {
                Ok(ControlFlow::Continue(()))
            }
        };

        match &filter {
            Some(shas) => walk_commits(repo_path, shas, &mut visit)?,
            None => walk_repo(repo_path, &mut visit)?,
        }
    }

    Ok(ExtractStats {
        count,
        limit_reached,
    })
}
